### ISSUE TITLE
pod with initcontainer failed with UnexpectedAdmissionError when restart kubelet

### ISSUE THREAD

**Issue by AllenXu93** [kubernetes/kubernetes#123971]
### What happened?

Memory manager have bug, restart kubelet (do not delete any checkpoint file ), pod with init-container will failed.

### What did you expect to happen?

Pod A and Pod B is still running.

### How can we reproduce it (as minimally and precisely as possible)?

I have a node had 256Gi and 2 NUMA node with 128Gi for one NUMA;
1. Create Pod A, with init-container and container both 100Gi
2. Create Pod B, with init-container and container both 100Gi 
3. Restart kubelet.
4. Both Pod A and Pod B failed with UnexpectedAdmissionError

### Anything else we need to know?

In `--topology-policy=single-numa-node` mode.
When memory manager caculate memory for topology, it will caculate init-container firstly, when comes to app container, it will reuse the init-container's memory as possible,  and reduce init-container's memory in state, it's as expected;

https://github.com/kubernetes/kubernetes/blob/9df98f3d393e204dd8db02322057ac2e7f3f6c28/pkg/kubelet/cm/memorymanager/policy_static.go#L192
https://github.com/kubernetes/kubernetes/blob/9df98f3d393e204dd8db02322057ac2e7f3f6c28/pkg/kubelet/cm/memorymanager/policy_static.go#L956-L965

But when restart kubelet, memory manager will re calculate memory for init-container,  if now node has not enough memory for init-container, the whole pod will be failed because of Admit failed;

<img width="1060" alt="image" src="https://github.com/kubernetes/kubernetes/assets/151703335/d455fd2e-b7a6-4661-825a-ae661d96d84a">

After restart kubelet:
<img width="1420" alt="image" src="https://github.com/kubernetes/kubernetes/assets/151703335/cd1f196d-b3c7-418f-b670-90bb51051321">

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.17", GitCommit:"a7736eaf34d823d7652415337ac0ad06db9167fc", GitTreeState:"clean", BuildDate:"2022-12-08T11:47:36Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.17", GitCommit:"a7736eaf34d823d7652415337ac0ad06db9167fc", GitTreeState:"clean", BuildDate:"2022-12-08T11:42:04Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by AllenXu93** [comment]
I think we need to fix this bug by change logic:  if init-container is already terminated, skip memory / cpu topology calculate;
And we still should not make any running pod failed after restart kubelet ;

**Comment by fredkan** [comment]
Why memory not enough, when kubelet restart? it calc main-container(100G) first? and left 28Gi for init-container?

**Comment by AllenXu93** [comment]
> Why memory not enough, when kubelet restart? it calc main-container(100G) first? and left 28Gi for init-container?

When kubelet restart, it will calc all the pods on node once again,  and containers in pod's order is calc init-container then app-container;

**Comment by likakuli** [comment]
is this exists in master branch?

**Comment by ffromani** [comment]
/triage accepted
/priority backlog

We seen similar issues in the past. I'll have a look and crosslink them. I think the issue is legit and there's something to fix here. Setting priority low pending further investigation

**Comment by ffromani** [comment]
There's also IIRC an issue about terminated pods re-doing admission on kubelet restart. I think we should not do admission for those containers, because, well, the container was already terminated and is not going to consume resources, but we need to be very careful assessing the implication of this change.

I'll try to find and link the relevant issues

somehow related: https://github.com/kubernetes/kubernetes/issues/117955

**Comment by AllenXu93** [comment]
> There's also IIRC an issue about terminated pods re-doing admission on kubelet restart. I think we should not do admission for those containers, because, well, the container was already terminated and is not going to consume resources, but we need to be very careful assessing the implication of this change.
> 
> I'll try to find and link the relevant issues
> 
> somehow related: #117955

Yes, admission on kubelet restart should skip for both terminated pods or terminated init-containers in pods.
In our case, restart kubelet cause pod Failed is a critical problem. Case we some times need to update kubelet config like change pidMax or reverse resources of Node, in this case pod failed will have bad impact on online services.

**Comment by sunziping2016** [comment]
Maybe PR https://github.com/kubernetes/kubernetes/pull/126343 is related.

**Comment by AllenXu93** [comment]
您好，邮件已收到，我会尽快给您回复。

**Comment by AllenXu93** [comment]
> Maybe PR #126343 is related. I've digged for hours for `UnexpectedAdmissionError`, and found that PR. I see a lot of issues reporting `UnexpectedAdmissionError` and maybe many of them are correlated.

PR https://github.com/kubernetes/kubernetes/pull/126343  will skip pod admit when pod is Terminated (Failed or Success).
In this issue,  running pod will be effected.

**Comment by sunziping2016** [comment]
> > Maybe PR #126343 is related. I've digged for hours for `UnexpectedAdmissionError`, and found that PR. I see a lot of issues reporting `UnexpectedAdmissionError` and maybe many of them are correlated.
> 
> PR https://github.com/kubernetes/kubernetes/pull/126343  will skip pod admit when pod is Terminated (Failed or Success).
> In this issue,  running pod will be effected.

Here is what I found: when kubelet restarts, it receives a list of Pods from apiserver, and tries to admit them one by one. It allocates CPU resources even if the pod is Terminated. Pods admitted later may not have enough CPU resources, and these pods are marked Failed with UnexpectedAdmissionError **no matter they are Running or Succeeded**. After all pods were processed, kubelet releases resources allocated to Terminated pods in removeStaleState(), so the issue won't happen again after admitting the initial list of pods.

**Comment by AllenXu93** [comment]
> > > Maybe PR #126343 is related. I've digged for hours for `UnexpectedAdmissionError`, and found that PR. I see a lot of issues reporting `UnexpectedAdmissionError` and maybe many of them are correlated.
> > 
> > 
> > PR #126343  will skip pod admit when pod is Terminated (Failed or Success).
> > In this issue,  running pod will be effected.
> 
> Here is what I found: when kubelet restarts, it receives a list of Pods from apiserver, and tries to admit them one by one. It allocates CPU resources even if the pod is Terminated. Pods admitted later may not have enough CPU resources, and these pods are marked Failed with UnexpectedAdmissionError **no matter they are Running or Succeeded**. After all pods were processed, kubelet releases resources allocated to Terminated pods in removeStaleState(), so the issue won't happen again after admitting the initial list of pods.

Every times before calc CPU resource, cpu-manager will call removeStaleState, to make sure pod that Terminated is clear in record. 
https://github.com/kubernetes/kubernetes/blob/85ff7e75a2066fef3543a55ca995c5173e779b72/pkg/kubelet/cm/cpumanager/cpu_manager.go#L328-L336

I'm not sure whether this can match your case;

**Comment by sunziping2016** [comment]
@AllenXu93 My case is that `removeStaleState()` only works after `sourcesReady.AllReady()`. This is most likely false when processing the initial list of Pods.
https://github.com/kubernetes/kubernetes/blob/e39571591d2917ca642f47ebca11253a91793c4b/pkg/kubelet/cm/cpumanager/cpu_manager.go#L362-L368

I think `removeStaleState()` is a legacy workaround. It has existed since (at least) v1.18. And the PR I referred can solve the issue completely. It's only cherry-picked to v1.28 and later . Anyway, I don't know whether your case is the same as my case. If you think your case is different, you can ignore my comments. Sorry for bothering you.

**Comment by AllenXu93** [comment]
> @AllenXu93 My case is that `removeStaleState()` only works after `sourcesReady.AllReady()`. This is most likely false when processing the initial list of Pods.
> 
> https://github.com/kubernetes/kubernetes/blob/e39571591d2917ca642f47ebca11253a91793c4b/pkg/kubelet/cm/cpumanager/cpu_manager.go#L362-L368
> 
> I think `removeStaleState()` is a legacy workaround. It has existed since (at least) v1.18. And the PR I referred can solve the issue completely. It's only cherry-picked to v1.28 and later . Anyway, I don't know whether your case is the same as my case. If you think your case is different, you can ignore my comments. Sorry for bothering you.

My case is different, in my case, running pods that have init-container will re calculate memory NUMA topo, which may cause the running pod to become failed if node's remain memory is not enough for init-container;

**Comment by AllenXu93** [comment]
您好，邮件已收到，我会尽快给您回复。
